presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -0,0 +1,554 @@
+            StreamPreferredProperties requiredProperties;
+            StreamPreferredProperties preferredProperties;
+            if (node.isPartial()) {
+                requiredProperties = parentPreferences.withoutPreference().withDefaultParallelism(session);
+                preferredProperties = parentPreferences.withDefaultParallelism(session);
+            }
+            else {
+                // a final changes the input organization completely, so we do not pass through parent preferences
+                requiredProperties = singleStream();
+                preferredProperties = defaultParallelism(session);
+            }
+
+            return planAndEnforceChildren(node, requiredProperties, preferredProperties);
+        }
+
+        @Override
+        public PlanWithProperties visitDistinctLimit(DistinctLimitNode node, StreamPreferredProperties parentPreferences)
+        {
+            // final limit requires that all data be in one stream
+            StreamPreferredProperties requiredProperties;
+            StreamPreferredProperties preferredProperties;
+            if (node.isPartial()) {
+                requiredProperties = parentPreferences.withoutPreference().withDefaultParallelism(session);
+                preferredProperties = parentPreferences.withDefaultParallelism(session);
+            }
+            else {
+                // a final changes the input organization completely, so we do not pass through parent preferences
+                requiredProperties = singleStream();
+                preferredProperties = defaultParallelism(session);
+            }
+
+            return planAndEnforceChildren(node, requiredProperties, preferredProperties);
+        }
+
+        @Override
+        public PlanWithProperties visitEnforceSingleRow(EnforceSingleRowNode node, StreamPreferredProperties parentPreferences)
+        {
+            return planAndEnforceChildren(node, singleStream(), defaultParallelism(session));
+        }
+
+        //
+        // Nodes that require parallel streams to be partitioned
+        //
+
+        @Override
+        public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferredProperties parentPreferences)
+        {
+            StreamPreferredProperties requiredProperties;
+            StreamPreferredProperties preferredChildProperties;
+
+            if (node.getStep() == Step.FINAL || node.getStep() == Step.SINGLE) {
+                // final aggregation requires that all data be partitioned
+                // todo split aggregation during enforcement
+                requiredProperties = parentPreferences.withDefaultParallelism(session).withPartitioning(node.getGroupBy());
+                preferredChildProperties = parentPreferences.withDefaultParallelism(session)
+                        .withPartitioning(node.getGroupBy())
+                        .constrainSymbols(node.getSource().getOutputSymbols());
```

Comment:
It means limit the perferences to only use the specified symbols.  So if you are partitioned on symbols not in that list, you need to drop the partitioning declaration.